### PR TITLE
fix(iso8601): fixes & unification & perf across all langs

### PR DESCRIPTION
**Pull request by ttodua**

### CODE REVIEW

File: ts/src/test/base/test.datetime.ts [hunk 1]
```diff
@@ -23,6 +23,54 @@ function testIso8601 () {
...
+    // plain-integer strings are accepted
+    assert (exchange.iso8601 ('1755432123456') === '2025-08-17T12:02:03.456Z');
+    // strings that are not a plain integer are rejected
+    assert (exchange.iso8601 ('123abc') === undefined);
+    // non-integer numbers are floored
+    assert (exchange.iso8601 (514862627559.9) === '1986-04-26T01:23:47.559Z');
+    // last representable millisecond of year 9999
+    assert (exchange.iso8601 (253402300799999) === '9999-12-31T23:59:59.999Z');
+    // one millisecond past the maximum supported range yields undefined
+    assert (exchange.iso8601 (8640000000000001) === undefined);
```
Comment (carlotestor) [inline ts/src/test/base/test.datetime.ts:73]: This suite now locks in the string-tightening and range guards, but the non-finite axis — the actual bug fixed in the Python, PHP and C# ports (NaN/Infinity/1e300 previously returned a bogus epoch string or raised) — has no coverage. NaN/Infinity literals are not portable across the transpiled languages, but `assert (exchange.iso8601 (1e300) === undefined)` is: it exercises the `> 8.64e15` guard in every port, including PHP's `is_finite` branch which is what rejects NaN. Worth one line.

File: go/v4/exchange_orderbook.go [hunk 2]
```diff
@@ -136,20 +148,28 @@ func (this *WsOrderBook) Limit() any {
...
 	}
 
-	if timestamp, ok := snapshot.(map[string]any)["timestamp"]; ok {
-		this.Timestamp = timestamp.(int64)
-		this.Datetime = Iso8601(timestamp.(int64))
+	if timestamp, ok := snapshotMap["timestamp"]; ok && timestamp != nil {
+		ts := ParseInt(timestamp)
+		if ts != math.MinInt64 {
+			this.Timestamp = &ts
+			this.Datetime = Iso8601(ts)
```
Comment (carlotestor) [inline go/v4/exchange_orderbook.go]: This is the fix for the master panic I reproduced: `Update()` asserted `timestamp.(int64)` on a snapshot that had been through `json.Unmarshal`, where every number is `float64`. The `ParseInt` normalization is the right shape — and note the `ts != math.MinInt64` guard keeps a truly unparsable timestamp from being recorded. Suggest a small regression case in the order-book shape test feeding a `json.Unmarshal`-decoded snapshot (float64 timestamp + nonce) so this cannot silently come back.

**Review by carlotestor — COMMENTED** [review]
## Summary
Cross-language `iso8601` hardening plus a Go WS order-book fix. The shared parity gate is `ts/src/test/base/test.datetime.ts`, which transpiles to all six languages, and every hand-written port (Python, PHP, C#, Go, Java) is touched in this PR.

## Verification (executed, not just read)
- **JS differential:** re-implemented `iso8601` checked against a `Date.toISOString()` oracle over **537,252,021** points (every day boundary across the full Date range, dense ms sweeps around epoch and now, hour boundaries, 3M random scatter, boundaries ±1) — **0 diffs**, including the extended-year `+275760-09-13T00:00:00.000Z` format.
- **Ports:** transpiled `test.datetime.ts` to every language and ran it — Python ✓, PHP ✓ (`zend.assertions=1`), Go ✓ (`go build ./tests/base/` + base REST suite), C# ✓ base REST suite under `UTC`, `America/New_York`, `Asia/Tokyo`, `Australia/Lord_Howe`, Java ✓ via a direct probe against the worktree sources covering all 48 new asserts plus the non-finite axis.
- **Go:** `go build ./v4/ ./v4/pro/` ✓, `go vet ./v4/` ✓, order-book shape tests pass.

## What this PR actually fixes (measured on master vs head)
- **NaN/Infinity (Python, PHP, C#):** master returned a bogus epoch (`1970-01-01T00:00:00.000Z`) from PHP/C# and raised from Python for NaN/Infinity/1e300; all now return null/undefined, matching JS. This is reachable in Python because `json.loads` accepts bare `NaN`/`Infinity`.
- **C# banker's rounding:** `Convert.ToInt64(514862627559.9)` rounded **up** to `…47.560Z` on master; the double + `Math.Floor` path now floors to `…47.559Z` like JS. The `InvariantCulture` on `ToString` also fixes wrong-year output under non-Gregorian host cultures (e.g. `th-TH`).
- **C# DateTimeKind.Unspecified → Utc** on the epoch constructor.
- **Go order book:** master reported `timestamp: 0` / `datetime: null` for snapshots **without** a timestamp (and for a genuine epoch-0 timestamp), and `WsOrderBook.Update()` **panicked** (`interface conversion: interface {} is float64, not int64`) on any snapshot decoded via `encoding/json`, since JSON numbers arrive as `float64`. The `*int64` field plus `ParseInt` normalization fixes all three; I reproduced the panic on master and confirmed it is gone on this head.

## Notes
- `WsOrderBook.Timestamp` is now `*int64` — a public API shape change for Go consumers. Justified (absent vs epoch-0 need to be distinguishable, and both `timestamp` and `datetime` were wrong on master), but worth a changelog call-out.
- Known deliberate divergence: Python rejects `bool` input while C# coerces `true` → 1 ms. Pre-existing, not exercised by any exchange path, and the test file's comment already scopes asserts to the 1970–9999 range where implementations agree.
- The added string-`/^[0-9]+$/` tightening and the 8.64e15 cap are consistent across all six ports (probed per-language, incl. numeric-string, empty, `'  12'`, `1e300`, max±1).

Merge gate: 🟢
Merge probability: 90%

**Review by carlotestor — COMMENTED** [review]
Follow-up on 8edb0571cd4efbf486622859c0027023b096aefd

Unreviewed delta vs last review (`76d94fc`): `d759f720` (address go comments) + this head.

`Update()` now only nonce-compares (ParseInt both sides, MinInt64 ignored) and delegates timestamp/datetime/nonce to `Reset()`. `Reset()` clears `Timestamp`/`Datetime` when the snapshot has no timestamp (JS `reset()` parity). `Iso8601` rejects float64/float32 NaN/±Inf and out-of-range magnitudes before `ParseInt`.

This head locks the two earlier notes: `assert (exchange.iso8601 (1e300) === undefined)` in `test.datetime.ts`, and `TestWsOrderBookUpdateJSONDecodedSnapshot` feeding a real `json.Unmarshal` snapshot (float64 timestamp/nonce, stale nonce ignored, missing timestamp clears).

Ran on this SHA: that new test PASS; `AddElementToObject(book, "timestamp", int64)` still populates `*int64`; `Iso8601(514862627559.9)` / NaN / Inf / `1e300` / missing-vs-epoch-0 `Reset` match.

Merge gate: 🟢

**Review by carlotestor — COMMENTED** [review]
Follow-up on 513619cdce474743f1fdb78652b4fe57625b8db2

This head is a plain merge of master into `fixes-perf-iso8601-2`; no commit of this PR's own is new. I verified that by blob hash — all ten PR files (`ts/src/base/functions/time.ts`, `ts/src/test/base/test.datetime.ts`, `go/v4/exchange_{time,orderbook,helpers}.go`, `go/v4/exchange_orderbook_shape_test.go`, `cs/ccxt/base/Exchange.Time.cs`, `php/Exchange.php`, `python/ccxt/base/exchange.py`, `java/.../base/Time.java`) are byte-identical to `8edb057`, which I already reviewed. Nothing to re-review on the branch side.

What the merge does change is the claim in the title. Master brought in the Rust port (#28627), so "all langs" is now seven, and the Rust base has its own hand-written `iso8601` in `rust/ccxt-base/src/exchange.rs` (~line 1019) that this PR does not touch:

```rust
Value::Float(f) => f as i64,
```

Read against the asserts this PR adds to the shared `test.datetime.ts` (which transpiles to Rust as `rust/tests/base/test.datetime.rs`):

- **NaN / ±Infinity → epoch.** Rust's `as i64` is a *saturating* cast and maps `NaN` to `0`, so `iso8601(NaN)` returns `1970-01-01T00:00:00.000Z` — exactly the bogus-epoch bug this PR fixes in PHP and C#. `+Inf` saturates to `i64::MAX` and lands in the `from_timestamp_millis` `None` branch, so that one is accidentally fine.
- **The 8.64e15 cap is missing.** `chrono::from_timestamp_millis` accepts roughly ±262000 years, so `iso8601(8640000000000001)` yields a string in Rust where line 73 of `test.datetime.ts` now asserts `undefined`.
- **String acceptance is looser than `/^[0-9]+$/`.** `s.parse::<i64>()` accepts a leading `+`, so `iso8601('+5')` returns a datetime where JS returns `undefined`.

The currently committed `rust/tests/base/test.datetime.rs` predates this PR (no `1e300` / `8640000000000001` asserts), so whether this surfaces as red CI depends on whether the Rust suite regenerates from `test.datetime.ts` on this run — I could not build Rust in this environment to settle it, so I am reporting the source-level divergence rather than a CI verdict. Worth checking the Rust job on this head before merge; if it does regenerate, the same three-line guard you already added to Go (`ParseInt` + finite check + magnitude cap) is the whole fix.

Nothing else in the merge touches this PR's surface.

Merge gate: 🟢

**Review by carlotestor — COMMENTED** [review]
Follow-up on 11867298b165301f1bafd5e21aa5f2dfefa5a3e2

Delta vs my last review (`513619c`) is one line: `restore rust` added to `build/cleanup.sh`, alongside the existing directory entries (`java/lib/src/main/java/io/github/ccxt/types`, `examples`), so the directory form is consistent with how `restore` is already used there. Since `rust/**` is a fully regenerated, gitignored tree, restoring it after a build run is the right call and it does not touch this PR's `iso8601` surface.

Nothing else changed; the ten source files are unchanged from the head I already verified.

One carry-over that this commit does not address (not a new finding, just still open): the Rust base `iso8601` in `rust/ccxt-base/src/exchange.rs` still saturates `NaN` to epoch and lacks the 8.64e15 cap, so it diverges from the asserts this PR adds to `test.datetime.ts` if the Rust suite regenerates.

Merge gate: 🟢

**Comment by carlotestor** [comment]
## Merge digest

Unifies `iso8601` across all six languages — plain-integer strings accepted, NaN/±Inf and out-of-range rejected — behind precomputed zero-pad tables, and makes Go's `WsOrderBook.Timestamp` nullable.

| File | Δ | What |
| --- | ---: | --- |
| `ts/src/base/functions/time.ts` | +53/−7 | `iso8601` pad tables, string input |
| `ts/src/test/base/test.datetime.ts` | +52/−0 | new `iso8601` edge cases |
| `go/v4/exchange_time.go`, `exchange_helpers.go` | +40/−13 | `Iso8601` guards, `AddElementToObject` ptr |
| `go/v4/exchange_orderbook.go` (+shape test) | +131/−28 | `WsOrderBook.Timestamp` → `*int64` |
| `php/Exchange.php`, `python/ccxt/base/exchange.py` | +31/−11 | `iso8601` finite/range guards |
| `cs/…/Exchange.Time.cs`, `java/…/Time.java`, `build/cleanup.sh` | +46/−10 | `Iso8601` parity, restore rust |

```mermaid
flowchart LR
  NewCountedOrderBook["NewCountedOrderBook"]
  Iso8601["Iso8601"]
  cleanup_sh["cleanup.sh"]
  func["func"]
  NewIndexedOrderBook["NewIndexedOrderBook"]
  AddElementToObject["AddElementToObject"]
  strOrNil["strOrNil"]
  NewWsOrderBook["NewWsOrderBook"]
```

`ts/src/base/functions/time.ts`

```ts
@@ -52,23 +52,69 @@
+const iso8601TwoDigits = [];
+for (let i = 0; i < 60; i++) {
+    iso8601TwoDigits.push ((i < 10) ? ('0' + i) : ('' + i));
+}
+const iso8601ThreeDigits = [];
+for (let i = 0; i < 1000; i++) {
+    iso8601ThreeDigits.push ((i < 10) ? ('00' + i) : ((i < 100) ? ('0' + i) : ('' + i)));
+}
-    } else {
+    } else if ((typeof timestamp === 'string') && timestamp.match (/^[0-9]+$/)) {
-    if (Number.isNaN (_timestampNumber) || _timestampNumber < 0) {
+    if ((_timestampNumber === undefined) || Number.isNaN (_timestampNumber) || _timestampNumber < 0) {
-    try {
```
